healvis/observatory.py

```diff
@@ -307,16 +332,51 @@ def _vis_calc(self, pcents, tinds, shell, vis_array, Nfin, beam_pol="pI"):
             memory_usage_GB = resource.getrusage(resource.RUSAGE_SELF).ru_maxrss / 1e6
             north = self.north_poles[tinds[count]] if haspoles else None
             za_arr, az_arr, pix = self.calc_azza(c_, north, return_inds=True)
-            beam_cube = self.beam.beam_val(az_arr, za_arr, self.freqs, pol=beam_pol)
+            if isinstance(self.beam, list):
+                # Adds another dimension to beam_cube: the baselines.
+                # Beams may be different for each antenna, and they are
+                # not power beams.
+                # Multiplies the beams for the 2 antennas in a baseline.
+
+                # Accumulate the antenna numbers
+                antennas = set()
+                for bi, baseline in enumerate(self.array):
+                    assert baseline.ant1 is not None and baseline.ant2 is not None,  \
+                            "Antenna number not set for baseline "+str(bi)
+                    antennas.add(baseline.ant1)
+                    antennas.add(baseline.ant2)
+                assert len(antennas) == len(self.beam), "Number of beams does not match number of antennas"
+
+                beam_cube = [ None for i in range(len(self.array)) ]
+                beam_val = [ None for i in range(len(antennas)) ]
+                for i in antennas:
+                    bv = self.external_beam_val(self.beam[i], az_arr, za_arr, self.freqs, pol=beam_pol)
+                    bv[np.argwhere(za_arr>np.pi/2)[:, 0], :] = 0    # Sources below horizon
+                    beam_val[i] = bv
+                for bi, bl in enumerate(self.array):
+                    # Multiply beam correction for the two antennas in each baseline
+                    beam_cube[bi] = beam_val[bl.ant1]*beam_val[bl.ant2]
```

Comment:
Shouldn't one be complex-conjugated? Please ignore if I'm wrong about that.

Reply:
Yes I should fix that. All the beams we've used so far return real values but in general they could be complex.